signac/contrib/filterparse.py

```diff
@@ -183,7 +200,7 @@ def _parse_single(key, value=None):
     """
     if value is None or value == "!":
         return key, {"$exists": True}
-    elif _is_json(value):
+    elif _is_json(value) or _is_list(value):
```

Comment:
I wonder if `_is_json` should be implemented as:
```python
    return (q[0] == "{" and q[-1] == "}") or (q[0] == "[" and q[-1] == "]")
```

This would have several advantages:
- It's more accurate to the meaning of "valid JSON" because `[0, 1, 2]` [is valid JSON](https://www.json.org/json-en.html) per the specification
- It prevents the invalid use of list-like keys (see `elif _is_json(key)` and the `ValueError` it raises).
- It is faster to check the individual character at the start and end rather than `str.startswith`. I find it just as readable.
https://github.com/bdice/python_microbenchmarks/blob/main/microbenchmarks/str_startswith_index.py

Then the `_is_list` method can be removed.

```suggestion
    elif _is_json(value):
```

The only other thing this needs is a test in [test_shell.py](https://github.com/glotzerlab/signac/blob/master/tests/test_shell.py).

Reply:
Ok, yes, I think the change to `_is_json` is a lot cleaner. I also think the `ValueError` message should be raised at the beginning of the method. I'll make these changes and add the test. Thanks!

Reply:
One important question here is whether or not parsing a filter is ever a performance bottleneck. @bdice can you think of a performance-critical code path you've benchmarked where parsing is even a blip? I can't, but I'm also not in the right mindset at the moment to really think through this.

If not, I would advocate for correctness over performance. @AlainKadar Instead of checking `_is_json`, I would get rid of that method entirely and simply use the `_parse_json` method, which is a try-except that will tell you whether a string is valid JSON. It will be slower, but it will be more accurate and bypass any concerns about whether we are correctly identifying JSON strings.

Reply:
I agree that would be better. I generously assumed that the current code structure was designed this way on purpose for some reason, but I think it’s probably a good idea to just parse the whole input to ensure its validity as Vyas suggested.

Reply:
Ok makes sense. But I should still keep the `_is_json` for testing whether the key is a JSON, at the beginning of the method?